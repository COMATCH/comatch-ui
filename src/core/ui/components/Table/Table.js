import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './Table.scss';

const propTypes = {
    /**
     * Data goes in the `<td>`
     */
    cellData: PropTypes.arrayOf(PropTypes.object),
    children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
    /**
     * Additional class names
     */
    className: PropTypes.string,
    /**
     * Data goes in the row with the `<th>`
     */
    headings: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.string),
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
    /**
     * Alternative styling
     */
    secondary: PropTypes.bool,
};

const defaultProps = {
    cellData: [],
    children: null,
    className: '',
    headings: null,
    secondary: false,
};

function renderHeadings(headings) {
    if (headings.length && typeof headings[0] === 'string') {
        return (
            <tr>
                {headings.map((heading, key) => <th key={key}>{heading}</th>)}
            </tr>
        );
    }

    return headings;
}

function renderCells(cellData, children) {
    if (cellData.length) {
        return cellData.map((object, trKey) => (
            <tr key={trKey}>
                {Object.keys(object).map((key, tdKey) => <td key={tdKey}>{object[key]}</td>)}
            </tr>
        ));
    }

    return children;
}

/**
 * A styled HTML `<table>` to present and organize data.
 */

export const Table = ({ cellData, children, className, headings, secondary }) => (
    <div className={classNames('Table', className, { primary: !secondary, secondary })}>
        <table className="Table__table">
            {headings && (
                <thead>
                    {renderHeadings(headings)}
                </thead>
            )}
            <tbody>{renderCells(cellData, children)}</tbody>
        </table>
    </div>
);

Table.propTypes = propTypes;

Table.defaultProps = defaultProps;
