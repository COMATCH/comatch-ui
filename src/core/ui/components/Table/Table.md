
## Table

From [`Table.js`](Table.js)

A styled HTML `<table>` to present and organize data.

prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**cellData** | `Array[]<Object>` | `[]` | :x: | Data goes in the `<td>`
**children** | `Union<Array[]<ReactNode> \| ReactNode>` | `null` | :x: | 
**className** | `String` | `''` | :x: | Additional class names
**headings** | `Union<Array[]<String> \| Array[]<ReactNode> \| ReactNode>` | `null` | :x: | Data goes in the row with the `<th>`
**secondary** | `Boolean` | `false` | :x: | Alternative styling



