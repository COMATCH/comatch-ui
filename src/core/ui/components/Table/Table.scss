@import '../../../styles/variables.scss';

.Table {
    overflow-x: auto;

    &__table {
        border-collapse: collapse;
        margin-bottom: 20px;
        max-width: 100%;
        width: 100%;

        th,
        td {
            text-align: left;
            font-size: $co-font-sm;
            border-bottom: 1px solid $co-color-mid-gray;
        }
    }

    // Different styles
    &.primary {
        th {
            padding: 10px 10px 5px;
        }
        td {
            padding: 15px 10px;
        }
    }

    &.secondary {
        th {
            padding: 10px;
        }
        td {
            padding: 10px 10px;
        }
        tr, td, th {
            border: 1px solid $co-color-mid-gray;
        }
    }

    tr.Link {
        transition: color $co-transition-fast ease-out;

        &:hover {
            color: $co-color-primary;
            cursor: pointer;
        }
    }
}
