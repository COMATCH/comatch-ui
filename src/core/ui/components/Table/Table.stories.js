/* eslint-disable import/no-unresolved */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { Table } from './Table';

const headingsStrings = ['City', 'Street', 'Country', 'Phone'];

const headings = [
    <tr key={0}>
        <th key={0}>City</th>
        <th key={1}>Street</th>
        <th key={2}>Country</th>
        <th key={3}>Phone</th>
    </tr>
];

const rows = [
    { city: 'Athens', street: 'Venizelou', country: 'GR', phone: '01529032810' },
    { city: 'Berlin', street: 'Freistr', country: 'DE', phone: '01529032810' },
    { city: 'New York', street: 'Nyc str', country: 'US', phone: '01529032810' },
    { city: 'Bern', street: 'Randomstr', country: 'CH', phone: '01529032810' },
    { city: 'Rome', street: 'Colstr', country: 'IT', phone: '01529032810' },
];

storiesOf('Table', module)
    .add('rendering headings and children as objects with primary style', () => (
        <Table headings={headings}>
            {rows.map(({ city, street, country, phone }, key) => (
                <tr key={key}>
                    <td>{city}</td>
                    <td>{street}</td>
                    <td>{country}</td>
                    <td>{phone}</td>
                </tr>
            ))}
        </Table>
    ))
    .add('rendering headings and cellData (serving as children) with primary style', () => (
        <Table headings={headingsStrings} cellData={rows} />
    ))
    .add('rendering a single heading and row with primary style', () => (
        <Table headings={<tr><th>Single heading</th></tr>}>
            <tr>
                <td>Athens</td>
            </tr>
        </Table>
    ))
    .add('without headings, with rows and primary style', () => <Table cellData={rows} />)
    .add('rendering headings and children as objects with secondary style', () => (
        <Table headings={headings} secondary>
            {rows.map(({ city, street, country, phone }, key) => (
                <tr key={key}>
                    <td>{city}</td>
                    <td>{street}</td>
                    <td>{country}</td>
                    <td>{phone}</td>
                </tr>
            ))}
        </Table>
    ))
    .add('rendering headings and cellData (serving as children) with secondary style', () => (
        <Table headings={headingsStrings} cellData={rows} secondary />
    ))
    .add('rendering a single heading and row with secondary style', () => (
        <Table headings={<th>Single heading</th>} secondary>
            <tr>
                <td>Athens</td>
            </tr>
        </Table>
    ))
    .add('without headings, with rows and secondary style', () => (
        <Table cellData={rows} secondary />
    ));
