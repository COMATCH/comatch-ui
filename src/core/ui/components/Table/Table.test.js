import React from 'react';
import renderer from 'react-test-renderer';
import { Table } from './Table';
import '../../../config/tests/setup';

describe('Table', () => {
    it('should render empty table correctly', () => {
        const table = renderer.create(
            <Table></Table>).toJSON();

        expect(table).toMatchSnapshot();
    });

    it('should render table with data correctly', () => {
        const className = 'arbitrary-class-name';
        const headings = [
            <tr key={0}>
                <th key={0}>City</th>
                <th key={1}>Street</th>
            </tr>
        ];

        const rows = [
            { city: 'Athens', street: 'Venizelou' },
            { city: 'Berlin', street: 'Freistr' }
        ];

        const table = renderer.create(
            <Table headings={headings} className={className}>
                {rows.map(({ city, street }, key) => (
                    <tr key={key}>
                        <td>{city}</td>
                        <td>{street}</td>
                    </tr>
                ))}
            </Table>).toJSON();
        expect(table).toMatchSnapshot();
    });

    it('should render table with secondary class correctly', () => {
        const table = renderer.create(
            <Table
                secondary={true}>
            </Table>).toJSON();

        expect(table).toMatchSnapshot();
    });

    it('should render table cellData correctly', () => {

        const cellData = [
            { city: 'Athens', street: 'Venizelou' },
            { city: 'Berlin', street: 'Freistr' }
        ];

        const table = renderer.create(
            <Table
                cellData={cellData}>
            </Table>).toJSON();
        expect(table).toMatchSnapshot();
    });

    it('should render table with heading array correctly', () => {

        const headings = [ 'City', 'Street' ];

        const table = renderer.create(
            <Table
                headings={headings}>
            </Table>).toJSON();
        expect(table).toMatchSnapshot();
    });
});
