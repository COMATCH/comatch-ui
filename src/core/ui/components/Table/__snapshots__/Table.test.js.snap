// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render empty table correctly 1`] = `
<div
  className="Table primary"
>
  <table
    className="Table__table"
  >
    <tbody />
  </table>
</div>
`;

exports[`Table should render table cellData correctly 1`] = `
<div
  className="Table primary"
>
  <table
    className="Table__table"
  >
    <tbody>
      <tr>
        <td>
          Athens
        </td>
        <td>
          Venizelou
        </td>
      </tr>
      <tr>
        <td>
          Berlin
        </td>
        <td>
          Freistr
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table should render table with data correctly 1`] = `
<div
  className="Table arbitrary-class-name primary"
>
  <table
    className="Table__table"
  >
    <thead>
      <tr>
        <th>
          City
        </th>
        <th>
          Street
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Athens
        </td>
        <td>
          Venizelou
        </td>
      </tr>
      <tr>
        <td>
          Berlin
        </td>
        <td>
          Freistr
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table should render table with heading array correctly 1`] = `
<div
  className="Table primary"
>
  <table
    className="Table__table"
  >
    <thead>
      <tr>
        <th>
          City
        </th>
        <th>
          Street
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`Table should render table with secondary class correctly 1`] = `
<div
  className="Table secondary"
>
  <table
    className="Table__table"
  >
    <tbody />
  </table>
</div>
`;
